Add turn actions

oliverlee opened the pull request.

oliverlee committed:
Add method and tests for playing cards

Co-Authored-By: Björn Andersson <andebjor@gmail.com>

oliverlee committed:
Add methods to change turn phases

Co-Authored-By: Mick van Gelderen <mickvangelderen@gmail.com>

mickvangelderen commented on src/dominion/player.rs:
```diff
@@ -54,14 +74,213 @@ impl Player {
+    }
+
+    pub fn end_turn(&mut self) -> Result<()> {
+        if let Some(TurnPhase::Buy { buy: _, worth: _ }) = self.phase {
```
```suggestion
        if let Some(TurnPhase::Buy { .. }) = self.phase {
```

mickvangelderen reviewed (commented):
Looks good overall and I think this should be merged since we shouldn't try to make it perfect until we know what we need. Some considerations:

- I expect most of the functions to be moved to the Arena struct since purchasing affects the stacks/piles and some cards will need put cards on other players' decks etc. 

- Consider renaming action and buy to remaining_actions, remaining_buys.

oliverlee commented:
Out and about so will address those points later and clean up the commit messages.

andebjor commented on src/dominion/player.rs:
```diff
@@ -54,14 +74,213 @@ impl Player {
             self.shuffle_deck();
         }
 
+        // TODO handle empty deck
```
/Isn't it so that if the pile is empty (after shuffling the discard pile) then nothing happens? If so, simply adding the check `if !self.deck_pile.is_empty() {` should be enough,

But perhaps the caller might wanna know if a draw actually happened. Then I guess returning a `bool` makes sense?

oliverlee replied:
Yeah I guess we'll figure out how to do this later. `remove` will panic if the desk is empty.

andebjor commented on src/dominion/player.rs:
```diff
@@ -54,14 +74,213 @@ impl Player {
     }
+
+    pub fn start_turn(&mut self) -> Result<()> {
+        if let None = self.phase {
```
Well, I don't know rust :) but this (`if let None = ...`) strikes me as very odd. I'm assuming setting `self.phase` to `Action`. But why do it like this?

oliverlee replied:
The idea is that a player should only be able to start a turn if that player's turn is not already active.

```rust
if let None = self.phase {
    // do stuff
}
```
So here we check if `self.phase` is `None` and if it is, then do stuff. If it is `Some`, we don't really care what is contained in the `self.phase` option and simply return an error.

andebjor replied:
Of course, I get it now. Thanks!

mickvangelderen replied:
There is a method on Option to do exactly this called `is_none`.

andebjor commented on src/dominion/player.rs:
```diff
@@ -54,14 +74,213 @@ impl Player {
+        } else if let Some(phase) = &mut self.phase {
+            match phase {
+                TurnPhase::Action { action, buy, worth } => {
+                    if action > &mut 0 {
```
Why is the `&mut` needed in front of the `0`?

oliverlee replied:
`action` is `&mut` so `0` needs to be the same type for comparison.

andebjor replied:
Indeed. I must say that I am surprised that one cannot compare a non-`mut` to a `mut`, but OK, that's how it is. Would be interested in knowing the rational, however :)

But this seems to work: `*action > 0`, Not sure which variant is more "rusty", but to me this one feels more sane than creating a `mut` reference to an integer constant just to be able to do a comparison.

mickvangelderen replied:
@andebjor I'd have to agree. It is clearer to dereference action which is `Copy` and do the comparison.

oliverlee committed:
Improve card and player names

Rename CardAction to CardEffect as 'action' is a type of Card and a
resource.

Rename TurnPhase enum fields
  from: action, buy, worth
  to: remaining_actions, remaining_buys, total_wealth
for more explicit descriptions of the resources.
